File: src/gigs/gigs.controller.ts
import { Controller, Get, Post, Put, Patch, Delete, Body, Param, UseGuards } from '@nestjs/common';
import { ApiTags, ApiBearerAuth } from '@nestjs/swagger';
import { GigsService } from './gigs.service';
import { JwtAuthGuard } from '../auth/guards/jwt.guard';
import { BandGuard } from '../common/guards/band.guard';
import { CurrentUser } from '../common/decorators/current-user.decorator';

@ApiTags('gigs')
@ApiBearerAuth('JWT')
@Controller('gigs')
@UseGuards(JwtAuthGuard, BandGuard)
export class GigsController {
  constructor(private readonly gigs: GigsService) {}

  // -- Gigs --

  @Get()
  findAll(@CurrentUser() user: any) { return this.gigs.findAll(user.bandId); }

  @Post()
  create(@CurrentUser() user: any, @Body() dto: any) { return this.gigs.create(user.bandId, dto); }

  @Put(':id')
  update(@CurrentUser() user: any, @Param('id') id: string, @Body() dto: any) {
    return this.gigs.update(user.bandId, id, dto);
  }

  @Patch(':id/status')
  updateStatus(@CurrentUser() user: any, @Param('id') id: string, @Body() dto: { status: string }) {
    return this.gigs.updateStatus(user.bandId, id, dto.status);
  }

  @Patch(':id/follow-up')
  updateFollowUp(@CurrentUser() user: any, @Param('id') id: string, @Body() dto: any) {
    return this.gigs.updateFollowUp(user.bandId, id, dto);
  }

  @Delete(':id')
  remove(@CurrentUser() user: any, @Param('id') id: string) { return this.gigs.remove(user.bandId, id); }

  // -- Contacts --

  @Get(':gigId/contacts')
  getContacts(@CurrentUser() user: any, @Param('gigId') gigId: string) {
    return this.gigs.getContacts(user.bandId, gigId);
  }

  @Post(':gigId/contacts')
  createContact(@CurrentUser() user: any, @Param('gigId') gigId: string, @Body() dto: any) {
    return this.gigs.createContact(user.bandId, gigId, dto);
  }

  @Delete(':gigId/contacts/:contactId')
  removeContact(@CurrentUser() user: any, @Param('gigId') gigId: string, @Param('contactId') contactId: string) {
    return this.gigs.removeContact(user.bandId, gigId, contactId);
  }

  // -- Checklists --

  @Get(':gigId/checklists')
  getChecklists(@CurrentUser() user: any, @Param('gigId') gigId: string) {
    return this.gigs.getChecklists(user.bandId, gigId);
  }

  @Post(':gigId/checklists')
  createChecklist(@CurrentUser() user: any, @Param('gigId') gigId: string, @Body() dto: { name: string }) {
    return this.gigs.createChecklist(user.bandId, gigId, dto);
  }

  @Delete(':gigId/checklists/:checklistId')
  removeChecklist(
    @CurrentUser() user: any,
    @Param('gigId') gigId: string,
    @Param('checklistId') checklistId: string,
  ) {
    return this.gigs.removeChecklist(user.bandId, gigId, checklistId);
  }

  // -- Checklist Items --

  @Get('checklists/:checklistId/items')
  getItems(@CurrentUser() user: any, @Param('checklistId') checklistId: string) {
    return this.gigs.getItems(user.bandId, checklistId);
  }

  @Post('checklists/:checklistId/items')
  createItem(@CurrentUser() user: any, @Param('checklistId') checklistId: string, @Body() dto: any) {
    return this.gigs.createItem(user.bandId, checklistId, dto);
  }

  @Put('checklists/:checklistId/items/:itemId')
  updateItem(
    @CurrentUser() user: any,
    @Param('checklistId') checklistId: string,
    @Param('itemId') itemId: string,
    @Body() dto: any,
  ) {
    return this.gigs.updateItem(user.bandId, checklistId, itemId, dto);
  }

  @Delete('checklists/:checklistId/items/:itemId')
  removeItem(
    @CurrentUser() user: any,
    @Param('checklistId') checklistId: string,
    @Param('itemId') itemId: string,
  ) {
    return this.gigs.removeItem(user.bandId, checklistId, itemId);
  }

  @Post('checklists/:checklistId/reset')
  resetItems(@CurrentUser() user: any, @Param('checklistId') checklistId: string) {
    return this.gigs.resetItems(user.bandId, checklistId);
  }

  // -- By-ID deletes --

  @Delete('contacts/:contactId')
  removeContactById(@CurrentUser() user: any, @Param('contactId') contactId: string) {
    return this.gigs.removeContactById(user.bandId, contactId);
  }

  @Delete('checklists/:checklistId')
  removeChecklistById(@CurrentUser() user: any, @Param('checklistId') checklistId: string) {
    return this.gigs.removeChecklistById(user.bandId, checklistId);
  }

  @Delete('checklist-items/:itemId')
  removeItemById(@CurrentUser() user: any, @Param('itemId') itemId: string) {
    return this.gigs.removeItemById(user.bandId, itemId);
  }
}

File: spring-src/main/java/com/blackout/api/gig/infrastructure/web/GigsController.java
package com.blackout.api.gig.infrastructure.web;

import com.blackout.api.gig.application.service.GigApplicationService;
import com.blackout.api.gig.infrastructure.web.dto.*;
import com.blackout.api.shared.domain.ForbiddenException;
import com.blackout.api.shared.infrastructure.security.BlackoutAuthentication;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/gigs")
public class GigsController {

    private final GigApplicationService service;

    public GigsController(GigApplicationService service) {
        this.service = service;
    }

    // ── Gigs ─────────────────────────────────────────────────────────────────

    @GetMapping
    public List<GigResponse> findAll(BlackoutAuthentication auth) {
        requireBand(auth);
        return service.findAll(auth.getBandId());
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public GigResponse create(@RequestBody CreateGigRequest req, BlackoutAuthentication auth) {
        requireBand(auth);
        return service.create(auth.getBandId(), req);
    }

    @PutMapping("/{id}")
    public GigResponse update(@PathVariable String id,
                               @RequestBody UpdateGigRequest req,
                               BlackoutAuthentication auth) {
        requireBand(auth);
        return service.update(auth.getBandId(), id, req);
    }

    @PutMapping("/{id}/status")
    public GigResponse updateStatus(@PathVariable String id,
                                     @RequestBody UpdateStatusRequest req,
                                     BlackoutAuthentication auth) {
        requireBand(auth);
        return service.updateStatus(auth.getBandId(), id, req.status());
    }

    @PutMapping("/{id}/follow-up")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void updateFollowUp(@PathVariable String id,
                                @RequestBody UpdateFollowUpRequest req,
                                BlackoutAuthentication auth) {
        requireBand(auth);
        service.updateFollowUp(auth.getBandId(), id, req);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void remove(@PathVariable String id, BlackoutAuthentication auth) {
        requireBand(auth);
        service.remove(auth.getBandId(), id);
    }

    // ── Summary — declared before /{id}/contacts to avoid ambiguity ──────────

    @GetMapping("/{id}/summary")
    public GigSummaryResponse getSummary(@PathVariable String id, BlackoutAuthentication auth) {
        requireBand(auth);
        return service.getSummary(auth.getBandId(), id);
    }

    // ── Contacts ─────────────────────────────────────────────────────────────

    @GetMapping("/{gigId}/contacts")
    public List<GigContactResponse> getContacts(@PathVariable String gigId,
                                                  BlackoutAuthentication auth) {
        requireBand(auth);
        return service.getContacts(auth.getBandId(), gigId);
    }

    @PostMapping("/{gigId}/contacts")
    @ResponseStatus(HttpStatus.CREATED)
    public GigContactResponse createContact(@PathVariable String gigId,
                                             @RequestBody CreateGigContactRequest req,
                                             BlackoutAuthentication auth) {
        requireBand(auth);
        return service.createContact(auth.getBandId(), gigId, req);
    }

    @DeleteMapping("/{gigId}/contacts/{contactId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void removeContact(@PathVariable String gigId,
                               @PathVariable String contactId,
                               BlackoutAuthentication auth) {
        requireBand(auth);
        service.removeContact(auth.getBandId(), gigId, contactId);
    }

    // ── Checklists ───────────────────────────────────────────────────────────

    @GetMapping("/{gigId}/checklists")
    public List<GigChecklistResponse> getChecklists(@PathVariable String gigId,
                                                     BlackoutAuthentication auth) {
        requireBand(auth);
        return service.getChecklists(auth.getBandId(), gigId);
    }

    @PostMapping("/{gigId}/checklists")
    @ResponseStatus(HttpStatus.CREATED)
    public GigChecklistResponse createChecklist(@PathVariable String gigId,
                                                 @RequestBody CreateChecklistRequest req,
                                                 BlackoutAuthentication auth) {
        requireBand(auth);
        return service.createChecklist(auth.getBandId(), gigId, req);
    }

    @DeleteMapping("/{gigId}/checklists/{checklistId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void removeChecklist(@PathVariable String gigId,
                                 @PathVariable String checklistId,
                                 BlackoutAuthentication auth) {
        requireBand(auth);
        service.removeChecklist(auth.getBandId(), gigId, checklistId);
    }

    // ── Checklist Items — static prefix /checklists must come before /{id} ───

    @GetMapping("/checklists/{checklistId}/items")
    public List<ChecklistItemResponse> getItems(@PathVariable String checklistId,
                                                 BlackoutAuthentication auth) {
        requireBand(auth);
        return service.getItems(auth.getBandId(), checklistId);
    }

    @PostMapping("/checklists/{checklistId}/items")
    @ResponseStatus(HttpStatus.CREATED)
    public ChecklistItemResponse createItem(@PathVariable String checklistId,
                                             @RequestBody CreateChecklistItemRequest req,
                                             BlackoutAuthentication auth) {
        requireBand(auth);
        return service.createItem(auth.getBandId(), checklistId, req);
    }

    @PutMapping("/checklists/{checklistId}/items/{itemId}")
    public ChecklistItemResponse updateItem(@PathVariable String checklistId,
                                             @PathVariable String itemId,
                                             @RequestBody UpdateChecklistItemRequest req,
                                             BlackoutAuthentication auth) {
        requireBand(auth);
        return service.updateItem(auth.getBandId(), checklistId, itemId, req);
    }

    @DeleteMapping("/checklists/{checklistId}/items/{itemId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void removeItem(@PathVariable String checklistId,
                            @PathVariable String itemId,
                            BlackoutAuthentication auth) {
        requireBand(auth);
        service.removeItem(auth.getBandId(), checklistId, itemId);
    }

    @PostMapping("/checklists/{checklistId}/reset")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void resetItems(@PathVariable String checklistId, BlackoutAuthentication auth) {
        requireBand(auth);
        service.resetItems(auth.getBandId(), checklistId);
    }

    // ── Helper ───────────────────────────────────────────────────────────────

    private void requireBand(BlackoutAuthentication auth) {
        if (auth.getBandId() == null) throw new ForbiddenException("Token sin banda seleccionada.");
    }
}
